### DISCUSSION TITLE
vcpkg installing tom example and CMAKE_BUILD_TYPE

### DISCUSSION THREAD

**Discussion by silverqx** [silverqx/TinyORM#37] [General]
I want to discuss it, I looked at it now and it's correct.

All these `if()` conditions that contain the `CMAKE_BUILD_TYPE` in the `TinyDeployment.cmake` will never fail because the `TINY_VCPKG` is always before, like `(TINY_VCPKG AND CMAKE_BUILD_TYPE STREQUAL "Debug")` so if the `TINY_VCPKG` is `false` the second expression `CMAKE_BUILD_TYPE` will not be considered.

The `TINY_VCPKG` is `true` only in the vcpkg port files [here](https://github.com/silverqx/TinyORM/blob/main/cmake/vcpkg/ports/tinyorm/portfile.cmake#L28) and the `vcpkg_cmake_configure()` guarantees that the build will be `Ninja` or `Unix Makefiles` single-configuration, this [`if()` condition](https://github.com/microsoft/vcpkg/blob/8eb57355a4ffb410a2e94c07b4dca2dffbee8e50/ports/vcpkg-cmake/vcpkg_cmake_configure.cmake#L76C1-L76C1) guarantees it.

So can't happen that the code will be invoked with the multi-config generator.

The `CONFIGURATIONS` option doesn't support gen-ex so I can't write something like this:

```cmake
install(TARGETS ${TomExample_target} 
    CONFIGURATIONS $<$<BOOL:${TINY_VCPKG>:Release>
    EXPORT TinyOrmTargets RUNTIME
)
```

What means I would need two `install(TARGETS)` commands one for the vcpkg and the other for everything else and that's pointless (even if it is more correct).

What do you think? @SchaichAlonso

**Comment by silverqx** [comment]
I would rather throw FATAL_ERROR if `TINY_VCPKG AND TINY_IS_MULTI_CONFIG` is `true` in [`TinyHelpers.cmake`](https://github.com/silverqx/TinyORM/blob/main/cmake/CommonModules/TinyHelpers.cmake#L268).

And I also forgot the `string(TOLOWER ${CMAKE_BUILD_TYPE} build_type)`.

**Comment by SchaichAlonso** [comment]
You can write

```
install(TARGETS ${TomExample_target} 
    CONFIGURATIONS Release
    EXPORT TinyOrmTargets RUNTIME
)
```
which defines it to be installed in Release mode, but doesn't say what to do in Debug mode, i.e. it won't do anything in debug mode, which I think is what you want to achieve

**Comment by silverqx** [comment]
Exactly, that's it, in this case it would look like this:

```cmake
if(TOM_EXAMPLE)
    if(TINY_VCPKG)
        install(TARGETS ${TomExample_target}
            CONFIGURATIONS Release
            EXPORT TinyOrmTargets RUNTIME
        )
    else()
        install(TARGETS ${TomExample_target} EXPORT TinyOrmTargets RUNTIME)
    endif()
endif()
```

But this is unnecessary, current code works well in all scenarios.

**Reply by SchaichAlonso** [reply]
The else branch installs it independent of the build type. If you want to simplify the statement, just install it independent of the vcpkg setting and have vcpkg ignore the debug variant:

```
if(TOM_EXAMPLE)
    install(TARGETS ${TomExample_target} EXPORT TinyOrmTargets RUNTIME)
endif()
```

More technical, [vcpkg_copy_tools](https://github.com/microsoft/vcpkg/blob/8eb57355a4ffb410a2e94c07b4dca2dffbee8e50/scripts/cmake/vcpkg_copy_tools.cmake#L17), which is used to get the tom-example binary into the tools path, looks for a binary with provided name in `${CURRENT_PACKAGES_DIR}/bin`, which is where the release variant was installed into, and copies it, before invoking [vcpkg_clean_executables_in_bin](https://github.com/microsoft/vcpkg/blob/8eb57355a4ffb410a2e94c07b4dca2dffbee8e50/scripts/cmake/vcpkg_copy_tools.cmake#L45) to remove the binary from [both the release and the debug](https://github.com/microsoft/vcpkg/blob/8eb57355a4ffb410a2e94c07b4dca2dffbee8e50/scripts/cmake/vcpkg_clean_executables_in_bin.cmake#L30) installation prefixes.

**Comment by silverqx** [comment]
Yeah, I know about this and I also remember why I decided this way, I decided to have the portfile as small as possible, and as you can see I'm installing it in the way that vcpkg is expecting. So I'm handling it in the TinyDeployment.cmake instead of in the portfile.

I also agree that this solution that you are proposing is nicer, smaller, and more clear but I will not rewrite it again, it works well. 

Also, look a few lines [down](https://github.com/silverqx/TinyORM/blob/main/cmake/Modules/TinyDeployment.cmake#L56) there is another `CMAKE_BUILD_TYPE`, why would I remove two `CMAKE_BUILD_TYPE` and leave there this one, I would have to rewrite it all to get rid of all three `CMAKE_BUILD_TYPE` and there is no reason for it.

Everything is fine and works superb, except there are three `CMAKE_BUILD_TYPE` 😂😂👌
